add appointment status enum

seantcanavan opened the pull request:
## Add appointment status enum

The appointment status values are not well documented unfortunately and span multiple formats (int/str). This enum gives as much flexibility as possible while giving callers the opportunity to call them by name rather than their short letter code.

I have no idea how y'all do enums internally but hopefully this isn't too far off base. 

## Breaking Changes

N/A

seantcanavan commented:
Apologies - can't tag reviewers so have to @dillonstreator @elliehastings manually. Thanks in advance!

dillonstreator commented on athenahealth/enum/appointment_status.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+func (as AppointmentStatus) String() string {
+	return appointmentStatusToString[as]
```
I think we can do away with `appointmentStatusToString` in favor of `string(as)`

seantcanavan replied:
I'm not sure how we tell if a regular string is a valid AppointmentStatus or not but I'll make this change

dillonstreator commented on athenahealth/enum/appointment_status.go:
```diff
@@ -0,0 +1,41 @@
+package enum
```
this should just be in the top level package in the appointments file

dillonstreator commented on athenahealth/enum/appointment_status.go:
```diff
@@ -0,0 +1,41 @@
+
+import "strconv"
+
+// AppointmentStatus is derived from https://docs.athenahealth.com/api/api-ref/appointment#Book-appointment
```
could you add the following to the comment as well
```go
// The athenaNet appointment status. There are several possible statuses.
// x=cancelled
// f=future (It can include appointments where were never checked in, even if the appointment date is in the past. It is up to a practice to cancel appointments as a no show when appropriate to do so.)
// o=open
// 2=checked in
// 3=checked out
// 4=charge entered (i.e. a past appointment)
```

dillonstreator commented on athenahealth/enum/appointment_status.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+func (as AppointmentStatus) Valid() bool {
+	return appointmentStatusToString[as] != ""
```
`_, ok := map[val]` would be the preferred way to do this but I believe we'd prefer to have a slice of `AppointmentStatus`'s and iterate to check if `as` is included.

seantcanavan replied:
I'll add them to a local variable in `Valid()`. I don't want to pollute the package autocomplete options with an array that just holds status values that are just used internally in one function

dillonstreator commented on athenahealth/enum/appointment_status.go:
```diff
@@ -0,0 +1,41 @@
+	CheckedIn     AppointmentStatus = "2"
+	CheckedOut    AppointmentStatus = "3"
+	Future        AppointmentStatus = "f"
+	Open          AppointmentStatus = "o"
```
Can you please prefix these with `AppointmentStatus` i.e. `AppointmentStatusOpen` to provide namespacing on these enums? This becomes especially important for when they're moved back into the top level `athenahealth` package

dillonstreator reviewed (commented):
Thanks for the PR, @seantcanavan!
Left a few comments

seantcanavan commented:
no worries @dillonstreator :smile: - addressed comments!

threw in a call to `Valid()` in the `String()` implementation so we can still return `""` if the enum instance is not a valid AppointmentStatus. Without it - `String()` just echoes the current string which makes error checking more difficult

dillonstreator commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	for _, currentAS := range appointmentStatuses {
+		if string(currentAS) == string(as) {
```
no need to type convert here as we can directly compare the `AppointmentStatus`'s
```suggestion
		if currentAS == as {
```

dillonstreator commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,57 @@
+	AppointmentStatusOpen          AppointmentStatus = "o"
+)
+
+func (as AppointmentStatus) Int() int {
```
Can you please share your use case for interacting with AppointmentStatus as an `int`? None of the existing values for appointment status in this repo reference it as an `int`.

seantcanavan replied:
it was more for convenience I guess since they do send numbers back (in string format). i can remove this though no sweat.

dillonstreator commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	return ""
+}
```
I think this method should be as simple as `return string(as)` since the caller is still required to validate the appointment status by checking if the string is empty. We're not saving anything in the way of interacting with this by doing these extra steps. Infact, we're adding extra steps in type converting for values that we know are valid i.e. storing after already having validated.

dillonstreator commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,57 @@
+		AppointmentStatusCheckedOut,
+		AppointmentStatusFuture,
+		AppointmentStatusOpen,
+	}
```
we generally like to define this list just under the enum/const definitions to improve co-location. If/when a new const is introduced, it's a little clearer to add the new const to the list.

seantcanavan replied:
I agree however every piece of code is in the same package in this repo so I wasn't sure how else to hide this array which wouldn't be used for anything except for validation. if it's outside the Valid() function it will pollute the `client.appointment...` namespace / autocompletion in people's IDEs. 

I will move it though no sweat.

dillonstreator replied:
it shouldn't pollute the namespace for clients as it shouldn't be exported

seantcanavan replied:
I meant for all internal development / users since Appointment... already has a ton of autocomplete options already

dillonstreator replied:
ah fair - addressed here https://github.com/eleanorhealth/go-athenahealth/pull/105#discussion_r1269697713

dillonstreator commented on athenahealth/appointments.go:
```diff
@@ -173,8 +173,8 @@ func (h *HTTPClient) ListBookedAppointments(ctx context.Context, opts *ListBooke
 
-		if len(opts.AppointmentStatus) > 0 {
-			q.Add("appointmentstatus", opts.AppointmentStatus)
+		if opts.AppointmentStatus.Valid() {
```
An invalid appointment status should result in an error rather than a no-op

dillonstreator reviewed (commented):
Few more comments - not sure how I missed the question around `int` use case in the first review

seantcanavan commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+func (as AppointmentStatus) Valid() bool {
+	for _, currentAS := range appointmentStatuses {
```
is there any reason to keep this as an array instead of a map as I had previously?

dillonstreator replied:
I think the main drawback of the map in its previous form was the redundancy/duplication of the string value.
I could see the map being valuable in quickly identifying `AppointmentStatus` validity by using the boolean returned from indexing the map. We'd prefer that the map's value be empty structs.
```go
var appointmentStatusMap = map[AppointmentStatus]struct{}{
	AppointmentStatusCancelled: {},
	AppointmentStatusChargeEntered: {},
	AppointmentStatusCheckedIn: {},
	AppointmentStatusCheckedOut: {},
	AppointmentStatusFuture: {},
	AppointmentStatusOpen: {},
}

func (as AppointmentStatus) Valid() bool {
	_, ok := appointmentStatusMap[as]
	return ok
}
```

dillonstreator replied:
regarding polluting the internal package global namespace, I'd say it's probably fine to enclose `appointmentStatusMap` inside `Valid` until a use case for moving it out arises.

dillonstreator commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,37 @@
+	AppointmentStatusOpen          AppointmentStatus = "o"
+)
+
+var appointmentStatuses = map[AppointmentStatus]bool{
+	AppointmentStatusCancelled:     true,
+	AppointmentStatusChargeEntered: true,
+	AppointmentStatusCheckedIn:     true,
+	AppointmentStatusCheckedOut:    true,
+	AppointmentStatusFuture:        true,
+	AppointmentStatusOpen:          true,
+}
```
Curious - why `bool` instead of an empty struct `struct{}`? `false` would be a valid value but shouldn't dictate any meaningful behavior.
Empty struct is zero allocation and seeing that we don't care about the value it would be more appropriate.
https://go.dev/play/p/s4yNIUEv6G_M
https://dave.cheney.net/2014/03/25/the-empty-struct

```suggestion
var appointmentStatuses = map[AppointmentStatus]struct{}{
	AppointmentStatusCancelled:     {},
	AppointmentStatusChargeEntered: {},
	AppointmentStatusCheckedIn:     {},
	AppointmentStatusCheckedOut:    {},
	AppointmentStatusFuture:        {},
	AppointmentStatusOpen:          {},
}
```

seantcanavan replied:
mostly because I had no idea this was a thing - this is awesome! thanks I was trying to replace a 'set' from java/typescript.

dillonstreator commented on athenahealth/appointment_status.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func (as AppointmentStatus) Valid() bool {
+	return appointmentStatuses[as]
```
With the change to using empty `struct{}` as the map value, please also use the idiomatic 'key exists' syntax `_, ok := map[key]`. `false` is not an appropriate value in this map and shouldn't affect this behavior.

dillonstreator reviewed (changes requested):
Really appreciate the PR - hopefully last couple comments before getting this one merged!
